File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/FilterTableConnection.java
package com.emogoth.android.phone.mimi.db;

import com.emogoth.android.phone.mimi.db.model.Filter;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

public class FilterTableConnection {
    public static Single<List<Filter>> fetchFilters() {
        return DatabaseUtils.fetchTable(Filter.class, Filter.TABLE_NAME);
    }

    public static Single<List<Filter>> fetchFilters(String boardName, String filterName) {
        return DatabaseUtils.fetchTable(Filter.class, Filter.TABLE_NAME, null, Filter.BOARD + "=? AND " + Filter.NAME + "=?", boardName, filterName);
    }

    public static Single<List<Filter>> fetchFiltersByBoard(String boardName) {
        return DatabaseUtils.fetchTable(Filter.class, Filter.TABLE_NAME, null, Filter.BOARD + "=?", boardName);
    }

    public static Single<List<Filter>> fetchFiltersByName(String filterName) {
        return DatabaseUtils.fetchTable(Filter.class, Filter.TABLE_NAME, null, Filter.NAME + "=?", filterName);
    }

    public static Flowable<Boolean> addFilter(String name, String filter, String boardName, boolean highlight) {
        Filter filterModel = new Filter();
        filterModel.name = name;
        filterModel.filter = filter;
        filterModel.board = boardName;
        filterModel.highlight = (byte) (highlight ? 1 : 0);

        return DatabaseUtils.insertOrUpdateRow(filterModel);
    }

    public static Flowable<Boolean> removeFilter(String name) {
        Filter filterModel = new Filter();
        filterModel.name = name;
        return DatabaseUtils.removeRow(filterModel);
    }
}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/FilterTableConnection.kt
package com.emogoth.android.phone.mimi.db

import com.emogoth.android.phone.mimi.db.MimiDatabase.Companion.getInstance
import com.emogoth.android.phone.mimi.db.models.Filter
import io.reactivex.Single

object FilterTableConnection {
    @JvmStatic
    fun fetchFilter(): Single<List<Filter>> {
        return getInstance()?.filters()?.getAll() ?: Single.just(emptyList())
    }

    @JvmStatic
    fun fetchFilter(boardName: String, filterName: String): Single<Filter> {
        return getInstance()?.filters()?.getFilter(boardName, filterName) ?: Single.just(Filter())
    }

    @JvmStatic
    fun fetchFiltersByBoard(boardName: String): Single<List<Filter>> {
        return getInstance()?.filters()?.getFiltersByBoard(boardName) ?: Single.just(emptyList())
    }

    @JvmStatic
    fun fetchFiltersByName(filterName: String): Single<List<Filter>> {
        return getInstance()?.filters()?.getFiltersByName(filterName) ?: Single.just(emptyList())
    }

    @JvmStatic
    fun addFilter(filter: Filter): Single<Boolean> {
        return Single.defer {
            getInstance()?.filters()?.upsert(filter) ?: return@defer Single.just(false)
            Single.just(true)
        }
    }

    @JvmStatic
    fun removeFilter(boardName: String, name: String): Single<Boolean> {
        return getInstance()?.filters()?.removeFilter(boardName, name)?.map { value: Int -> value > 0 }
                ?: Single.just(false)
    }
}